test/e2e/framework/framework.go

```diff
@@ -393,7 +394,14 @@ func (f *Framework) AfterEach() {
 					} else {
 						Logf("Namespace %v was already deleted", ns.Name)
 					}
+				} else {
```

Comment:
So... why not just delete the namespaces yourself in a defer in your test?  Defer should still get a chance to run on ctrl+c

Reply:
If ctrl+c isn't allowing defers in the main body of a test to run we have a bigger problem.

Reply:
When ginkgo runner receives `ctrl-c` - it starts executing `afterSuite` - https://github.com/onsi/ginkgo/blob/master/internal/specrunner/spec_runner.go#L237 and exits immediately. If we don't cause process to wait in `afterSuite` (i.e what waiting for namespace to be deleted causes) then e2e process will exit without running defer or `aftereach`.

Reply:
`ctrl-c` does not allow defers in the main body of test to run.  

```
C
---------------------------------------------------------
Received interrupt.  Running AfterSuite...
^C again to terminate immediately
Apr  8 11:23:24.313: INFO: Running AfterSuite actions on all nodes
Apr  8 11:23:24.313: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "csi-mock-volumes-9650" for this suite.
Apr  8 11:23:24.317: INFO: Running AfterSuite actions on node 1
{"msg":"Test Suite completed","total":6,"completed":0,"skipped":486,"failed":0}

Ran 1 of 487 Specs in 8.019 seconds
FAIL! -- 1 Passed | 0 Failed | 0 Pending | 486 Skipped
```

As you can see cleanup function mentioned in defer() ^ did not run.

Reply:
I tried to solve this a long time ago and failed.  Ginkgo needs to be cut out like cancer.

Reply:
I've been playing with this a bit. SIGINT is handled by ginkgo, it does not call any defers or anything. A  test continue doing whatever it was doing before SIGINT.

In a separate goroutine, ginkgo runs AfterSuite, that deletes all namespaces, sets `f.ClientSet` to nil and call `os.Exit`. All this while the test is still testing! `os.Exit` won't call any defers either.

With this PR, AfterSuite waits a bit for the namespaces to be deleted. During this time, the test has at least chance to notice that something's wrong (e.g. it can't create new pods) and "fail" in a natural way (`framework.ExpectNoError`), which calls defers and `AfterEach`.

Still, the test namespaces are Terminating, which means Kubernetes will kill everything there in a random order. This is not exactly what the test needs - it may want to delete some stuff earlier than other, that's why we have `defers` and `AfterEach`. At least CSI tests suffers from this, CSI driver pod(s) must be deleted *after* a pod and PVC that uses a volume provided by the CSI driver. Kubelet can't unmount a volume, detach it or delete it without the driver still running.

Even with this drawback, I think it will help cleaning up some tests.

Reply:
@jsafrane

> With this PR, AfterSuite waits a bit for the namespaces to be deleted. During this time, the test has at least chance to notice that something's wrong (e.g. it can't create new pods) and "fail" in a natural way (framework.ExpectNoError), which calls defers and AfterEach.

IMO having WaitForNamespacesDeleted here is fine.
related https://github.com/kubernetes/kubernetes/issues/85541

just noting that for a ctrl+c on a finalizer test we might face https://github.com/kubernetes/kubernetes/issues/70916
so having the timeout of 5 minutes (as is in this PR) is a good idea.

> In a separate goroutine, ginkgo runs AfterSuite, that deletes all namespaces, sets f.ClientSet to nil and call os.Exit. All this while the test is still testing! os.Exit won't call any defers either.

that particular os.Exit call in ginkgo is bad, so we have to think of ways of avoiding it while we still need work to get done.

@ gnufied 
> Ginkgo has a weird bug that - AfterEach does not get called when
testsuite exits with certain kind of interrupt (Ctrl-C for example).
More info - onsi/ginkgo#222

my general question here would be why do we need to care about interupts outside of local / debugging scenarios?

cc @pohly @oomichi